File: 95 Unique Binary Search Trees II.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @paramn n: An integer
     * @return: A list of root
     */
    public List<TreeNode> generateTrees(int n) {
        if (n == 0) {
            return new ArrayList<TreeNode>();
        }
        return generate(1, n);
    }

    private ArrayList<TreeNode> generate(int s, int t) {
        ArrayList<TreeNode> nodeList = new ArrayList<TreeNode>();
        if (s > t) {
            nodeList.add(null);
            return nodeList;
        }
        for (int i = s; i <= t; ++i) {
            ArrayList<TreeNode> leftNodeList = generate(s, i - 1);
            ArrayList<TreeNode> rightNodeList = generate(i + 1, t);
            for (TreeNode leftNode : leftNodeList) {
                for (TreeNode rightNode : rightNodeList) {
                    TreeNode node = new TreeNode(i);
                    node.left = leftNode;
                    node.right = rightNode;
                    nodeList.add(node);
                }
            }
        }
        return nodeList;
    }
}

File: 95 Unique Binary Search Trees II.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  vector<TreeNode*> generateTrees(int n) {
    if (n == 0) {
      vector<TreeNode*> nodes;
      return nodes;
    }
    return generateRange(1, n);
  }

private:
  vector<TreeNode*> generateRange(int start, int end) {
    vector<TreeNode*> nodes;
    if (start > end) {
      nodes.push_back(NULL);
    } else {
      for (int i = start; i <= end; ++i) {
        vector<TreeNode*> left_tree = generateRange(start, i - 1);
        vector<TreeNode*> right_tree = generateRange(i + 1, end);
        for (int j = 0; j < static_cast<int>(left_tree.size()); ++j) {
          for (int k = 0; k < static_cast<int>(right_tree.size()); ++k) {
            TreeNode *node = new TreeNode(i);
            node->left = left_tree.at(j);
            node->right = right_tree.at(k);
            nodes.push_back(node);
          }
        }
      }
    }
    return nodes;
  }
};
